File: tests/fixtures/owasp/redirect.py
from flask import Flask, redirect, request

app = Flask(__name__)


@app.route("/login/callback")
def login_callback():
    return_url = request.args.get("return_url")  # tainted
    return redirect(return_url)  # CWE-601 sink


@app.route("/goto")
def goto():
    url = request.args.get("url")  # tainted
    return redirect(url)  # CWE-601 sink


@app.route("/auth/next", methods=["POST"])
def auth_next():
    next_url = request.form["next"]  # tainted
    return redirect(next_url)  # CWE-601 sink


# SAFE: startswith check for relative path
@app.route("/safe-redirect")
def safe_redirect():
    return_url = request.args.get("return_url", "/")
    if not return_url.startswith("/"):  # sanitizer
        return "invalid redirect", 400
    return redirect(return_url)


# SAFE: hardcoded redirect
@app.route("/home")
def home():
    return redirect("/dashboard")

File: legacy/tests/pre-pivot/fixtures/owasp/redirect.ts
import express from "express";

const app = express();

// VULNERABLE: open redirect from query param
app.get("/login/callback", (req, res) => {
  const returnUrl = req.query.returnUrl; // tainted
  res.redirect(returnUrl as string); // CWE-601 sink
});

// VULNERABLE: open redirect from body
app.post("/auth/complete", (req, res) => {
  const next = req.body.next; // tainted
  res.redirect(next); // CWE-601 sink
});

// VULNERABLE: Location header from user input
app.get("/goto", (req, res) => {
  const url = req.query.url; // tainted
  res.setHeader("Location", url as string); // CWE-601 sink
  res.status(302).end();
});

// SAFE: redirect to relative path only (startsWith check)
app.get("/safe-redirect", (req, res) => {
  const returnUrl = req.query.returnUrl as string;
  if (!returnUrl.startsWith("/")) { // sanitizer
    return res.status(400).send("invalid redirect");
  }
  res.redirect(returnUrl);
});

// SAFE: redirect to hardcoded URL
app.get("/home", (req, res) => {
  res.redirect("/dashboard");
});
